geneticalgorythm/algorithm.py

```diff
@@ -0,0 +1,119 @@
+        self.initial_attributes_standard_deviation = initial_attributes_standard_deviation
+        self.min_attributes_count = min_attributes_count
+        self.population_size = population_size
+        self.classifier = classifier
+        self.fitness_function = (self.classifier.calculate_precision_recall
+                                 if fitness_function == FitnessFunction.AveragePrecision
+                                 else self.classifier.calculate_area_under_roc)
+
+    def _generate_single_individual(self) -> np.ndarray:
+        features_count = int(np.random.normal(self.min_attributes_count, self.initial_attributes_standard_deviation, 1))
+        features = set(random_features_indices(features_count))
+
+        return binary_representation_of_features(features)
+
+    def _generate_initial_population(self) -> Population:
+        return [self._generate_single_individual() for _ in range(self.population_size)]
+
+    def _compute_fitness(self, population: Population) -> PopulationWithFitness:
+        # fitness function expect ordered LIST of features but we are passing a SET (witch has its own order).
+        # I think that it doesn't matter but when we have whole algorithm we need to test it.
+        return [(self.fitness_function(set_representation_of_features(individual)), individual) for individual in
+                population]
+
+    def _selection(self, population_with_fitness: PopulationWithFitness) -> List[Tuple[int, int]]:
+        return [(0, 0) for _ in range(len(population_with_fitness))]
+
+    def _cross(self, mother: np.ndarray, father: np.ndarray) -> np.ndarray:
+        loci = sorted(random_features_indices(self.loci_count - 1))
+        mother_parts = np.array_split(mother, loci)
+        father_parts = np.array_split(father, loci)
+
+        # [(0,1), (2, 3), (4, 5)...]
+        consecutive_pairs = ((even, odd) for even, odd in
+                             zip(range(0, self.loci_count, 2), range(1, self.loci_count, 2)))
+        parts = [(mother_parts[even], father_parts[odd]) for even, odd in consecutive_pairs]
+
+        return np.concatenate(list(itertools.chain(*parts)))
+
+    def _crossover(self, selected_pairs: List[Tuple[int, int]], population: Population) -> Population:
+        return [self._cross(population[mother], population[father]) for mother, father in selected_pairs]
+
+    def _mutate(self, individual: np.ndarray) -> None:
+        features_to_mutate = random_features_indices(
+            int(Algorithm.ALL_ATTRIBUTES_COUNT * self.chromosomes_to_mutate_coefficient))
+
+        for feature in features_to_mutate:
+            individual[feature] = 1 if individual[feature] == 0 else 0
+
+    def _mutation(self, population: Population) -> Population:
+        individuals_to_mutate = random.sample(range(len(population)),
+                                              int(len(population) * self.individuals_to_mutate_coefficient))
```

Comment:
konieczny tutaj jest cast na inta?

Reply:
z tego co pamietam to random.sample nie przyjmował floating point numbers, bo ta funkcja losuje unikalne elementy w liczbie podanej w drugim argumencie - liczba musi być intem